File: apps/frontend/src/lib/conversations.ts
/**
 * The conversation-state brick: one owner for "which thread am I in, what threads do I
 * have, and what did they say". The sidebar and the chat view are siblings that both need
 * it, so it lives here rather than in either of them (CLAUDE.md, one concern one module).
 *
 * A thread is created lazily, when the first question titles it: a thread posted on the New
 * chat click would sit in the rail under the placeholder title until it was asked something.
 * New chat therefore opens an empty draft and `startThread` registers it the moment there is
 * a question.
 *
 * `titleThread` is the second half of that (ADR 0012 as amended): once a turn is over, the
 * server generates the few-word label from the thread's exchanges and this store adopts the row
 * it answers with. Adopting the response rather than re-listing is deliberate - the PATCH body IS
 * the stored row, so one request settles it and the rail cannot reorder or flicker around the
 * thread the reader is looking at. A failed refresh is logged and nothing else: the thread
 * keeps the title it already had, which is a title, not an error to report.
 *
 * It runs more than once now (issue #118): the chat view asks again after each of the thread's
 * first `title_turns` turns, so a thread that opened with a greeting ends up named after the
 * question that followed, and the rail re-renders from this store either way. When the window
 * closes, or when the reader has renamed the thread themselves, the server answers with the row
 * unchanged - which is why this store can stay a plain "ask and adopt" and hold no rule of its
 * own about which name wins.
 *
 * `rename` is the reader's own half of the same request: the PATCH carries the title they typed
 * and the server stamps the row as theirs for good. This store holds no rule about that either -
 * it sends the words and adopts the row, exactly as `titleThread` does, so the two can never
 * disagree here about which name the rail shows.
 *
 * `replay` holds what the server remembers of the open thread, already folded into turns by
 * `lib/trace.ts`: the questions, the answers, and the whole trace each turn produced - its
 * reasoning, its calls and their arguments, its SQL pairs, tables and charts, its retries and its
 * refusals (ADR 0012 as amended, issue #90). The store folds it here, once, and through the same
 * fold the live stream goes through, so the chat view receives turns of the same shape whether
 * they were streamed or reopened. `chatKey` changes on every switch: it is the signal the chat
 * view resets its live turns on.
 *
 * The thread's `in_flight` goes into that fold (issue #143). A turn runs to its end on the server
 * whether or not anyone is watching, so a thread can be reopened while one is still going: the
 * server says so, and the fold renders that last question as still being answered rather than as
 * a turn whose trace was lost.
 *
 * `select` answers with the thread it opened, or null when the registry would not hand it over -
 * a deleted thread and another identity's are the same 404 and get the same answer here. The
 * shell restoring a thread named in the URL is the caller that needs to tell the two apart, and
 * this store is the only place that knows (issue #135, `lib/location.ts`).
 */

import { useCallback, useEffect, useState } from "react";

import {
  ApiError,
  createConversation,
  deleteConversation,
  getConversation,
  listConversations,
  renameConversation,
  retitleConversation,
} from "./api";
import type { Thread } from "./api";
import { replayTurns } from "./trace";
import type { Turn } from "./trace";

const LIST_FAILURE = "Could not load your conversations.";
const OPEN_FAILURE = "Could not open that conversation.";
const DELETE_FAILURE = "Could not delete that conversation.";
const RENAME_FAILURE = "Could not rename that conversation.";
const TITLE_FAILURE = "the conversation title was not refreshed";

export interface ConversationsStore {
  threads: Thread[];
  activeId: string | null;
  /** The open thread's past turns, as the server still remembers them. */
  replay: Turn[];
  /** Bumped on every thread switch; the chat view drops its live turns when it changes. */
  chatKey: number;
  loading: boolean;
  error: string | null;
  newChat: () => void;
  /** Opens a thread and resolves with the id that ended up open, or null when it could not be. */
  select: (threadId: string) => Promise<string | null>;
  remove: (threadId: string) => void;
  /** Names a thread as the reader typed it and adopts the row the server answers with. */
  rename: (threadId: string, title: string) => void;
  /** Registers the thread the first question of a draft belongs to and returns its id. */
  startThread: (title: string) => Promise<string>;
  /** Has the server title the thread from the turn it just had, and adopts the row it stores. */
  titleThread: (threadId: string) => void;
}

interface Open {
  activeId: string | null;
  replay: Turn[];
  chatKey: number;
}

const DRAFT: Open = { activeId: null, replay: [], chatKey: 0 };

/** The threads of whoever is signed in: loaded on mount, so a re-login lists that identity's. */
export function useConversations(): ConversationsStore {
  const [threads, setThreads] = useState<Thread[]>([]);
  const [open, setOpen] = useState<Open>(DRAFT);
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState<string | null>(null);

  useEffect(() => {
    let live = true;
    listConversations()
      .then((list) => {
        if (live) setThreads(list);
      })
      .catch((cause) => {
        if (live) setError(reason(cause, LIST_FAILURE));
      })
      .finally(() => {
        if (live) setLoading(false);
      });
    return () => {
      live = false;
    };
  }, []);

  const newChat = useCallback(() => {
    setError(null);
    setOpen((previous) => ({ ...DRAFT, chatKey: previous.chatKey + 1 }));
  }, []);

  // Reopening the thread already open is a no-op: it would drop the traces of this session's turns.
  const select = useCallback(
    (threadId: string) => {
      if (threadId === open.activeId) return Promise.resolve<string | null>(threadId);
      setError(null);
      setLoading(true);
      return getConversation(threadId)
        .then((conversation) => {
          setOpen((previous) => ({
            activeId: conversation.thread_id,
            replay: replayTurns(
              conversation.messages,
              conversation.turns,
              conversation.in_flight,
            ),
            chatKey: previous.chatKey + 1,
          }));
          return conversation.thread_id;
        })
        .catch((cause) => {
          setError(reason(cause, OPEN_FAILURE));
          return null;
        })
        .finally(() => setLoading(false));
    },
    [open.activeId],
  );

  const remove = useCallback((threadId: string) => {
    setError(null);
    deleteConversation(threadId)
      .then(() => {
        setThreads((previous) => previous.filter((thread) => thread.thread_id !== threadId));
        setOpen((previous) =>
          previous.activeId === threadId
            ? { ...DRAFT, chatKey: previous.chatKey + 1 }
            : previous,
        );
      })
      .catch((cause) => setError(reason(cause, DELETE_FAILURE)));
  }, []);

  const rename = useCallback((threadId: string, title: string) => {
    setError(null);
    renameConversation(threadId, title)
      .then((renamed) =>
        setThreads((previous) =>
          previous.map((thread) => (thread.thread_id === renamed.thread_id ? renamed : thread)),
        ),
      )
      .catch((cause) => setError(reason(cause, RENAME_FAILURE)));
  }, []);

  const startThread = useCallback(async (title: string) => {
    const thread = await createConversation(title);
    setThreads((previous) => [thread, ...previous]);
    setOpen((previous) => ({ ...previous, activeId: thread.thread_id }));
    return thread.thread_id;
  }, []);

  const titleThread = useCallback((threadId: string) => {
    retitleConversation(threadId)
      .then((titled) =>
        setThreads((previous) =>
          previous.map((thread) => (thread.thread_id === titled.thread_id ? titled : thread)),
        ),
      )
      .catch((cause) => console.warn(`secure-rls: ${reason(cause, TITLE_FAILURE)}`));
  }, []);

  return {
    threads,
    activeId: open.activeId,
    replay: open.replay,
    chatKey: open.chatKey,
    loading,
    error,
    newChat,
    select,
    remove,
    rename,
    startThread,
    titleThread,
  };
}

function reason(cause: unknown, fallback: string): string {
  return cause instanceof ApiError ? cause.message : fallback;
}

File: apps/backend/conversations.py
"""Conversation registry brick (ADR 0012) - the fifth tenant-scoped data path.

Every thread row carries the `sub` and `tenant_id` of the identity that created it, and
every read, write and delete is filtered by BOTH. A thread that belongs to another
identity raises the same `NotFound` with the same message as a thread that never existed,
so the API cannot be used to probe for foreign thread ids (existence non-disclosure).
`rename_thread` is scoped by the same clause as the rest: a rename aimed at another
identity's thread changes nothing and answers exactly like a missing one.

Titles are normalized on every write, whoever wrote them (ADR 0012 as amended): control and
formatting characters become spaces, whitespace collapses, and the result is cut to the
configured cap. A generated title is model output, so the store is the last place that can
guarantee what the sidebar renders is one line of displayable text - no NUL, no escape
sequence, no bidi override reordering the rail.

Turn history (ADR 0012 as amended, issue #90). Beside the thread rows this module keeps the whole
turn: the trace events it produced, in order - the model's reasoning per round, every tool call
with the arguments the model wrote, each call's one outcome (its result payload, the retry that
fed an error back, or the security refusal that ended it), and the terminal frame with the turn's
status, its telemetry and the prompt-guardrail position that produced it. A reopened thread
therefore replays as the conversation that happened, not as a tidied answer, which is the whole
auditability claim of this product.

One row per turn, `events` a JSON array of what history keeps, `cut` how many pieces the caps
refused. What reduces a live event stream to that row is `turns.py`, which owns the caps as well;
this module stores what it is handed and never inspects it. Rows are scoped by the same
`sub` + `tenant_id` clause as the thread rows, on both the read and the write, and only the newest
`conversations.max_history_turns` turns of a thread keep their history - an older turn replays as
text, as it did before any of this was stored. Deleting a thread deletes its history with it.

A read never repairs: a row whose JSON cannot be parsed raises rather than replaying a partial
turn as if it were whole. The lenient half is the write (`turns.py`), where the answer has already
streamed and losing the history is not worth failing a good turn over.

Storage: this module owns its own SQLite file (`state.db`, path injected by the caller).
That is a documented exception to the "only db.py opens a connection" rule - the registry
is application state, not tenant DATA, and it never touches the employee tables that the
scoped executor guards.

Dependencies: standard library only. LangGraph checkpointer rows for a deleted thread are
dropped by an optional `cleanup` callback the caller supplies, so nothing here imports
langgraph and the registry stays testable without it.
"""

import json
import sqlite3
import unicodedata
import uuid
from collections.abc import Callable
from contextlib import closing
from dataclasses import dataclass
from datetime import UTC, datetime
from pathlib import Path

from auth import Identity
from runtime import runtime

_THREADS_SCHEMA = """
CREATE TABLE IF NOT EXISTS threads (
    thread_id TEXT PRIMARY KEY,
    sub       TEXT NOT NULL,
    tenant_id TEXT NOT NULL,
    title     TEXT NOT NULL,
    created   TEXT NOT NULL
)
"""

_TURN_HISTORY_SCHEMA = """
CREATE TABLE IF NOT EXISTS turn_history (
    thread_id TEXT NOT NULL,
    sub       TEXT NOT NULL,
    tenant_id TEXT NOT NULL,
    turn      INTEGER NOT NULL,
    events    TEXT NOT NULL,
    cut       INTEGER NOT NULL,
    created   TEXT NOT NULL,
    PRIMARY KEY (thread_id, turn)
)
"""

# Discards the payloads issue #70 stored: turn history replaced them, no code reads them any more,
# and leaving tenant-derived rows on disk unreachable is worse (owner-approved, issue #90).
_SUPERSEDED_SCHEMA = "DROP TABLE IF EXISTS tool_results"

_NOT_FOUND_MESSAGE = "conversation not found"
# Cc and Cf: NUL and escape sequences, and the bidi overrides that could reorder the rail.
_CONTROL_CATEGORIES = frozenset({"Cc", "Cf"})


class NotFound(Exception):
    """Raised identically for a missing thread and for another identity's thread."""


@dataclass(frozen=True)
class Thread:
    """A conversation as the API returns it; the owning identity stays server-side."""

    thread_id: str
    title: str
    created: str


@dataclass(frozen=True)
class TurnHistory:
    """One past turn as the store keeps it: its trace events in order, and what the caps refused.

    `turn` is the ordinal of the question that opened the turn, counted from one, so a replayed
    transcript can put each turn's history back above the answer it produced. `events` are ADR
    0012 trace events, keyed exactly as the live stream keys them, so one fold renders a replayed
    turn and a live one. `cut` is how many pieces of the turn the caps refused, which is what lets
    a replay state that it is partial instead of reading as whole.
    """

    turn: int
    events: list[dict[str, object]]
    cut: int


class ConversationRegistry:
    """Tenant-scoped CRUD over the thread registry in its own SQLite state file."""

    def __init__(
        self,
        state_db: str | Path,
        clock: Callable[[], datetime] = lambda: datetime.now(UTC),
    ) -> None:
        """Open (creating if needed) the state file and pin the clock used for timestamps."""
        self._state_db = Path(state_db)
        self._clock = clock
        with closing(sqlite3.connect(self._state_db)) as conn:
            conn.execute(_THREADS_SCHEMA)
            conn.execute(_TURN_HISTORY_SCHEMA)
            conn.execute(_SUPERSEDED_SCHEMA)
            conn.commit()

    def create_thread(self, identity: Identity, title: str) -> Thread:
        """Register a new thread owned by this identity, with the title truncated to the cap."""
        thread = Thread(
            thread_id=uuid.uuid4().hex,
            title=_normalize_title(title),
            created=self._clock().isoformat(),
        )
        with closing(sqlite3.connect(self._state_db)) as conn:
            conn.execute(
                "INSERT INTO threads (thread_id, sub, tenant_id, title, created) "
                "VALUES (?, ?, ?, ?, ?)",
                (thread.thread_id, identity.sub, identity.tenant_id, thread.title, thread.created),
            )
            conn.commit()
        return thread

    def list_threads(self, identity: Identity) -> list[Thread]:
        """Return this identity's threads, newest first; another identity's are never visible."""
        with closing(sqlite3.connect(self._state_db)) as conn:
            rows = conn.execute(
                "SELECT thread_id, title, created FROM threads "
                "WHERE sub = ? AND tenant_id = ? ORDER BY created DESC, rowid DESC",
                (identity.sub, identity.tenant_id),
            ).fetchall()
        return [Thread(*row) for row in rows]

    def get_thread(self, identity: Identity, thread_id: str) -> Thread:
        """Return the identity's own thread, or raise NotFound for missing and foreign alike."""
        with closing(sqlite3.connect(self._state_db)) as conn:
            row = conn.execute(
                "SELECT thread_id, title, created FROM threads "
                "WHERE thread_id = ? AND sub = ? AND tenant_id = ?",
                (thread_id, identity.sub, identity.tenant_id),
            ).fetchone()
        if row is None:
            raise NotFound(_NOT_FOUND_MESSAGE)
        return Thread(*row)

    def rename_thread(self, identity: Identity, thread_id: str, title: str) -> Thread:
        """Retitle the identity's own thread; a foreign or missing id raises the same NotFound."""
        with closing(sqlite3.connect(self._state_db)) as conn:
            renamed = conn.execute(
                "UPDATE threads SET title = ? WHERE thread_id = ? AND sub = ? AND tenant_id = ?",
                (_normalize_title(title), thread_id, identity.sub, identity.tenant_id),
            ).rowcount
            conn.commit()
        if not renamed:
            raise NotFound(_NOT_FOUND_MESSAGE)
        return self.get_thread(identity, thread_id)

    def record_turn(self, identity: Identity, thread_id: str, history: TurnHistory) -> None:
        """Store one turn's history under the identity's own thread, pruning the older turns.

        The write is scoped by the same clause as every read, so a thread this identity does not
        own - or one deleted while its turn was still streaming - is written nothing and raises
        nothing: a foreign write is as indistinguishable from a missing thread as a foreign read
        is, and the caller is a stream teardown that has no reader left to tell.

        The turn is one row and is rewritten as a whole, so recording it twice cannot leave half of
        an older attempt behind, and the turns older than the newest `max_history_turns` are
        dropped as the new one lands - a long thread keeps its recent history, not all of it.
        """
        if not history.events:
            return
        scope = (thread_id, identity.sub, identity.tenant_id)
        with closing(sqlite3.connect(self._state_db)) as conn:
            if not self._owns(conn, identity, thread_id):
                return
            conn.execute(
                "INSERT OR REPLACE INTO turn_history "
                "(thread_id, sub, tenant_id, turn, events, cut, created) "
                "VALUES (?, ?, ?, ?, ?, ?, ?)",
                (
                    *scope,
                    history.turn,
                    json.dumps(history.events),
                    history.cut,
                    self._clock().isoformat(),
                ),
            )
            conn.execute(
                "DELETE FROM turn_history "
                "WHERE thread_id = ? AND sub = ? AND tenant_id = ? AND turn <= ?",
                (*scope, history.turn - runtime().conversations.max_history_turns),
            )
            conn.commit()

    def thread_turns(self, identity: Identity, thread_id: str) -> list[TurnHistory]:
        """Replay the stored turns of the identity's own thread, oldest first.

        Scoped like every other read: another identity's thread reads as empty, which is exactly
        what a thread that was never chatted in reads as. Nothing here decides whether the thread
        may be read at all - the caller asks the registry for it first, as it does for the
        transcript. A row that cannot be parsed raises: a replay that cannot be reconstructed says
        so rather than rendering a partial turn as a whole one.
        """
        with closing(sqlite3.connect(self._state_db)) as conn:
            rows = conn.execute(
                "SELECT turn, events, cut FROM turn_history "
                "WHERE thread_id = ? AND sub = ? AND tenant_id = ? ORDER BY turn",
                (thread_id, identity.sub, identity.tenant_id),
            ).fetchall()
        return [TurnHistory(turn, json.loads(events), cut) for turn, events, cut in rows]

    def delete_thread(
        self,
        identity: Identity,
        thread_id: str,
        cleanup: Callable[[str], None] | None = None,
    ) -> None:
        """Delete the identity's own thread and its stored history, then let `cleanup` run.

        `cleanup` drops the thread's checkpointer rows; the turn history is this store's own
        and goes in the same transaction as the thread row, under the same scoping clause.
        """
        with closing(sqlite3.connect(self._state_db)) as conn:
            deleted = conn.execute(
                "DELETE FROM threads WHERE thread_id = ? AND sub = ? AND tenant_id = ?",
                (thread_id, identity.sub, identity.tenant_id),
            ).rowcount
            conn.execute(
                "DELETE FROM turn_history WHERE thread_id = ? AND sub = ? AND tenant_id = ?",
                (thread_id, identity.sub, identity.tenant_id),
            )
            conn.commit()
        if not deleted:
            raise NotFound(_NOT_FOUND_MESSAGE)
        if cleanup is not None:
            cleanup(thread_id)

    def _owns(self, conn: sqlite3.Connection, identity: Identity, thread_id: str) -> bool:
        """Whether this identity's own thread exists, by the clause every access is scoped by."""
        row = conn.execute(
            "SELECT 1 FROM threads WHERE thread_id = ? AND sub = ? AND tenant_id = ?",
            (thread_id, identity.sub, identity.tenant_id),
        ).fetchone()
        return row is not None


def plain_one_line(text: str) -> str:
    """One line of displayable text: control and formatting characters out, whitespace collapsed.

    The shape a title has to be in before anything renders it, wherever the text came from. It
    is public because the titler needs the same judgment on a model's answer before it can tell
    a label from noise (`titles.py`), and two copies of "what is displayable" would drift.
    """
    displayable = "".join(
        " " if unicodedata.category(char) in _CONTROL_CATEGORIES else char for char in text
    )
    return " ".join(displayable.split())


def _normalize_title(title: str) -> str:
    """The title as the registry stores it: displayable, one line, cut to the configured cap."""
    return plain_one_line(title)[: runtime().conversations.title_max_chars]
